cram/cram_decode.c

```diff
@@ -1215,25 +1222,26 @@ static int cram_decode_seq(cram_fd *fd, cram_container *c, cram_slice *s,
                     if (ref_pos + pos-seq_pos > s->ref_end)
                         goto beyond_slice;
                     if (decode_md || decode_nm) {
-                        int i;
-                        for (i = 0; i < pos - seq_pos; i++) {
-                            // FIXME: not N, but nt16 lookup == 15?
-                            char base = s->ref[ref_pos - s->ref_start + 1 + i];
-                            if (base == 'N') {
-                                if (add_md_char(s, decode_md,
-                                                s->ref[ref_pos - s->ref_start + 1 + i],
+                        int i, nm_start = nm;
+                        const char *refp = s->ref + ref_pos - s->ref_start + 1;
+                        if (!memchr(refp, 'N', pos-seq_pos)) {
+                            md_dist += pos-seq_pos;
+                        } else {
+                            for (i = 0; i < pos - seq_pos; i++) {
+                                if (refp[i] != 'N')
+                                    continue;
+
+                                if (add_md_char(s, decode_md, refp[i],
```

Comment:
Doesn't this go wrong if there are two 'N's in range?  The old code would increment `md_dist` between them, but in this version the update is outside the inner loop so the second would always get `md_dist = 0`.

Reply:
The old code either incremented `nm` or `md_dist`, so the number of `md_dist` increments is equal to length - nm increment.  Thus I optimised it out by counting up one only.

Reply:
It does however look like the `seq[seq_pos-1+i] = base;` line got lost, hmm.

Reply:
The problem with `md_dist` is that [`add_md_char()`](https://github.com/samtools/htslib/blob/dcd20d9fab/cram/cram_decode.c#L1081) needs it, and before it was being incremented inside the `for (i = 0; i < pos - seq_pos; i++) {}` loop, but now it isn't.  So I don't see how `add_md_char()` could be getting the same value passed in as it used to?